bug(self-host): `cli_main.sfn` emits unmangled `compile_to_sailfin` when `build/native/.build-stamp` is absent

## Goal
Diagnose and fix the self-hosting regression where `sfn build -p compiler` (run by the freshly-built compiler binary against a clean cache) emits an **unmangled** `compile_to_sailfin` reference in `cli_main.sfn`'s object file, causing the link to fail — but only when `build/native/.build-stamp` is missing.

## Reproduction

```bash
ulimit -v 8388608
make clean-build
# Delete the Makefile stamp shim (or otherwise prevent the stamp from being written):
rm -f build/native/.build-stamp
make compile               # seed-build the new compiler; no stamp written

WORK=$(mktemp -d)
build/native/sailfin build -p compiler --work-dir "$WORK"
# Expected: clean self-host (0 hits / N misses, exit 0)
# Actual:   link failure
#   /usr/bin/ld: cli_main.sfn:(.text+0x6216): undefined reference to `compile_to_sailfin'
#   clang: error: linker command failed with exit code 1
#   link failed (clang exit=1)
```

Now write a stamp and re-try (no other changes):

```bash
echo "0.7.0-alpha.5+dev.testhash" > build/native/.build-stamp
WORK=$(mktemp -d)
build/native/sailfin build -p compiler --work-dir "$WORK"
# Expected and actual: built: <work>/native/sailfin
```

## What's strange
- `resolve_compiler_version` is only consumed in `cache_key_for` (cache key hashing), `_emit_capsule_artifact_sidecar`, `_emit_build_report`, and the `--version` print site. None should affect LLVM symbol mangling.
- `apply_module_symbol_mangling` (in `compiler/src/llvm/lowering/lowering_helpers_mangling.sfn`) ignores the compiler version entirely.
- Yet the **emitted `cli_main.o`** has an unmangled `compile_to_sailfin` reference when no stamp is present, and a properly-mangled `compile_to_sailfin__main` reference when a stamp is present.
- The repro is 0 cache hits / N misses in both cases — fresh compilation either way — so this isn't a stale-cache reuse problem in the obvious sense.

The bug was masked for a long time because `Makefile`'s rebuild target always wrote `build/native/.build-stamp` (via the shell shim that #516 retired). Issue #516's Phase 0.3 helper writes the same file from inside the compiler, but the helper only runs once `/pin-seed` advances past #516 (tracked by #757). In the interim, #516 keeps a transition bridge in the Makefile that re-writes the stamp the same way the old shim did — explicitly because removing it surfaces THIS bug.

## Scope
**In:**
- Reproduce the failure on a stamp-less, freshly-`clean-build`-ed tree.
- Diff the `cli_main.o` (or the upstream `.ll`) between stamp-present and stamp-absent cases to identify which compilation step diverges.
- Trace why `apply_module_symbol_mangling` (or its caller) takes a different code path based on what `resolve_compiler_version` returns.
- Fix the root cause so the stamp file's presence is irrelevant to compilation correctness — only `--version` reporting and cache key hashing should consult it.
- Drop the `[ ! -f build/native/.build-stamp ]` guard added in #516's Makefile bridge — once this bug is fixed AND #757 closes (seed bumped past the helper), the bridge becomes dead code.

**Out:**
- Re-organizing `resolve_compiler_version` / `resolve_compiler_version_for_cache` — the two-function split is intentional and orthogonal.
- Anything in `build_stamp.sfn` itself — the helper is correct; this bug predates and is independent of it.
- Compiler-version cache-key changes — the version is legitimately part of the cache hash.

## Acceptance Criteria
- [ ] Root cause identified in a comment on this issue (which call site, why it depends on `resolve_compiler_version`, and what the correct behaviour is).
- [ ] Fix lands so the four-state truth table from #516 holds even with no stamp file present:
  - clean tree, no stamp → self-host succeeds, version reports `<capsule_version>`
  - clean tree, stamp present → self-host succeeds, version reports stamp string
- [ ] `make clean-build && rm -f build/native/.build-stamp && make compile && WORK=$(mktemp -d) && build/native/sailfin build -p compiler --work-dir "$WORK"` exits 0.
- [ ] Once #757 closes too, the `[ ! -f build/native/.build-stamp ]` guard + transition bridge in `Makefile`'s rebuild target are removed in a follow-up.

## Files Affected
- Likely `compiler/src/llvm/lowering/lowering_helpers_mangling.sfn` or `compiler/src/llvm/lowering/entrypoints.sfn` — the actual divergence site needs to be located.
- Possibly `compiler/src/capsule_resolver.sfn` (around `resolve_compiler_version_for_cache` call at line 1694) if cache-key encoding leaks into compilation output.
- `Makefile` — drop the transition bridge once fixed AND #757 closes.

## Verification
```bash
ulimit -v 8388608
make clean-build
rm -f build/native/.build-stamp
make compile
WORK=$(mktemp -d)
timeout 300 build/native/sailfin build -p compiler --work-dir "$WORK"
test -x "$WORK/native/sailfin" && echo PASS || echo FAIL
rm -rf "$WORK"
```

## Size
M — investigation is the hard part. The fix itself may be one or two lines once the root cause is known. Diff `cli_main.o` (or the upstream `.ll`) between the two states to pinpoint where the divergence happens; that should make the call site obvious.

## Type
Bug

## Blocked by
None — independent of #516 and #757, just exposed by them.

## Context
This is the load-bearing reason #516's Makefile bridge can't be deleted yet. Once this bug is fixed AND #757 closes, both the bridge and this issue go away.

--- a/Makefile
+++ b/Makefile
@@ -700,13 +700,6 @@ rebuild:
 	@# recipe line. The `&&` chain ensures every diagnostic
 	@# message reaches the user before we exit.
 	@rm -f build/sailfin/program build/sailfin/program.ll
-	@# Invalidate any stale stamp from a prior rebuild so the
-	@# helper (post-#757) or the transition bridge below
-	@# always writes a fresh stamp reflecting the current git
-	@# HEAD + dirty state. Without this, `make rebuild` after a
-	@# new commit would leave the previous hash on disk and
-	@# `sfn --version` would report the wrong provenance.
-	@rm -f build/native/.build-stamp
 	@seed=$$(cat build/.seed-resolved); \
 	echo "[rebuild] running sfn build -p compiler (seed=$$seed)..."; \
 	build_rc=0; \
@@ -911,52 +904,6 @@ rebuild:
 			[ -f "$$manifest_path" ] || touch "$$manifest_path"; \
 		fi; \
 	done
-	@# Build stamp: epic #513 phase 0.3 / issue #516 moved this into
-	@# `compiler/src/build_stamp.sfn::emit_compiler_build_stamp_if_applicable`,
-	@# called from `cli_main.sfn`'s `is_build` post-link path. The
-	@# block below is a transition bridge — once `/pin-seed` advances
-	@# past #516 (tracked by #757) the helper writes the stamp during
-	@# `sfn build -p compiler` and the `[ ! -f ... ]` guard makes
-	@# this whole block a no-op. Delete only after BOTH #757 closes
-	@# (seed has the helper) AND #758 closes (self-host miscompilation
-	@# when `build/native/.build-stamp` is absent) — the bridge masks
-	@# #758 by guaranteeing the stamp always exists.
-	@if [ ! -f build/native/.build-stamp ]; then \
-		cap_version=$$(sed -n 's/^version *= *"\([^"]*\)"/\1/p' compiler/capsule.toml); \
-		if [ -z "$$cap_version" ]; then \
-			echo "[rebuild][warn] could not extract version from compiler/capsule.toml; skipping build stamp" >&2; \
-		else \
-			git_tag=$$(git describe --exact-match --tags HEAD 2>/dev/null || true); \
-			if [ -n "$$git_tag" ]; then \
-				build_version="$$cap_version"; \
-			else \
-				git_hash=$$(git rev-parse --short HEAD 2>/dev/null || true); \
-				git_dirty=""; \
-				if [ -n "$$git_hash" ]; then \
-					if ! git diff --quiet HEAD -- 2>/dev/null; then \
-						git_dirty=".dirty"; \
-					fi; \
-					build_version="$${cap_version}+dev.$${git_hash}$${git_dirty}"; \
-				else \
-					build_version="$${cap_version}+dev"; \
-				fi; \
-			fi; \
-			echo "$$build_version" > build/native/.build-stamp; \
-			echo "[rebuild] build stamp (transition bridge): $$build_version"; \
-		fi; \
-	fi
-	@# Cache invalidation companion to the bridge above: the seed
-	@# (pre-#757) populated `build/cache/` while resolving
-	@# `compiler_version` from `compiler/capsule.toml` (no stamp file
-	@# existed yet — we deleted it at the top of `rebuild`). With
-	@# the bridge now writing the stamp, downstream `sfn build -p
-	@# compiler` invocations (e.g. `test_work_dir_parity.sh`) resolve
-	@# the SAME version string and hit the seed's cached `cli_main.o`
-	@# — which carries the unmangled-symbol miscompilation tracked
-	@# by #758. Wiping the cache forces those downstream callers to
-	@# re-compile through the freshly-built (correct) compiler.
-	@# Delete with the rest of the bridge once #757 + #758 close.
-	@rm -rf build/cache
 	@echo "[rebuild] built $(NATIVE_OUT)"
 
 # =============================================================================

